toolkit/tools/scheduler/schedulerutils/initializegraph.go

```diff
@@ -17,18 +18,30 @@ const (
 )
 
 // InitializeGraph initializes and prepares a graph dot file for building.
-// - It will load and return the graph.
-// - It will subgraph the graph to only contain the desired packages if possible.
-func InitializeGraph(inputFile string, packagesToBuild []*pkgjson.PackageVer, deltaBuild bool) (isOptimized bool, pkgGraph *pkggraph.PkgGraph, goalNode *pkggraph.PkgNode, err error) {
+//   - It will use an existing graph if provided (graph should not have been previously initialized, input file must be "")
+//   - It will load from disk if a path is provided (existing graph must be nil)
+//   - It will subgraph the graph to only contain the desired packages if possible.
+//   - If canUseCachedImplicit is true, it will use cached nodes to resolve implicit dependencies instead of waiting for
+//     them to be built in the graph (This can allow the graph to be optimized immediately instead of waiting for the
+//     implicit nodes to be resolved by an unknown package later in the build).
+func InitializeGraph(inputFile string, existingGraph *pkggraph.PkgGraph, packagesToBuild []*pkgjson.PackageVer, canUseCachedImplicit bool) (isOptimized bool, pkgGraph *pkggraph.PkgGraph, goalNode *pkggraph.PkgNode, err error) {
 	const (
 		strictGoalNode = true
 	)
-	// Delta builds can use cached implicit nodes
-	canUseCachedImplicit := deltaBuild
+	timestamp.StartEvent("graph initialization", nil)
+	defer timestamp.StopEvent(nil)
 
-	pkgGraph, err = pkggraph.ReadDOTGraphFile(inputFile)
-	if err != nil {
-		return
+	if existingGraph == nil {
+		pkgGraph, err = pkggraph.ReadDOTGraphFile(inputFile)
+		if err != nil {
+			return
+		}
+	} else {
+		if len(inputFile) != 0 {
+			err = fmt.Errorf("input file '%s' specified along with existing graph, use one or the other", inputFile)
+			return
+		}
+		pkgGraph = existingGraph
```

Comment:
>pkgGraph = existingGraph [](http://example.com/codeflow?start=2&length=24)

I'm not sure what this is for, yet, but seeing how we're using pointers, we're effectively modifying `existingGraph`. For such cases, could we instead extract all of the code below this line into a new public function and call that whenever we need to update/reset an existing graph? And retain the single responisibility of the original `InitializeGraph`?

Reply:
Another bonus is that we won't need to pass `nil` if we don't need the `existingGraph` argument.

Reply:
Ok, I think I've split these in a meaningful way. We now have `InitializeGraphFromFile()` which calls `PrepareGraphForBuild()`.